Add FreeBind=true to sshd.socket example

sublimino opened the pull request:
Update recommendation as per https://github.com/coreos/bugs/issues/57#issuecomment-55455975 otherwise the socket will fail to start if the network isn't up yet. On my test hosts this happens regularly, which requires a machine reboot to try the race again :)

robszumski commented:
@marineam Can you review this?

marineam commented:
I am baffled as to why this is required since it is going to be binding to `0.0.0.0` or `::`. Either something else is going on or my understanding is flawed. Will review further...

sublimino commented:
With the example currently in the docs (without this proposed change) I get a race condition at startup:

    Jan 24 06:33:16 coreos1 systemd[1]: sshd.socket failed to listen on sockets: Address already in use
    Jan 24 06:33:16 coreos1 systemd[1]: Unit sshd.socket entered failed state.
    Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Calling unit command "restart" on "sshd.socket"'
    Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Result of "restart" on "sshd.socket": failed

From the [Arch docs](https://wiki.archlinux.org/index.php/Secure_Shell#Managing_the_sshd_daemon):

> Warning: Using sshd.socket effectively negates the ListenAddress setting, so using the default sshd.socket will allow connections over any address. To achieve the effect of setting ListenAddress, you must create a custom unit file and modify ListenStream (i.e. ListenStream=192.168.1.100:22 is equivalent to ListenAddress 192.168.1.100). You must also add FreeBind=true under [Socket] or else setting the IP address will have the same drawback as setting ListenAddress: the socket will fail to start if the network is not up in time.

crawford commented:
That is inline with the [systemd docs](http://www.freedesktop.org/software/systemd/man/systemd.socket.html#FreeBind=):
>  Controls whether the socket can be bound to non-local IP addresses. This is useful to configure sockets listening on specific IP addresses before those IP addresses are successfully configured on a network interface. This sets the IP_FREEBIND socket option. For robustness reasons it is recommended to use this option whenever you bind a socket to a specific IP address.

But we aren't binding to a specific address. I wonder if there is a race triggered by the restart command.

sublimino commented:
Let me know if you want entire startup logs and cloudconfig, or anything else.

crawford commented:
I think I know what's going on. Our example is not a runtime unit. On first boot, everything works fine. sshd.socket listens on 22. Cloud-init applies the config and restarts sshd.socket, moving it to 2222. Fine. On the second boot, the altered socket unit is already in place, so sshd.socket is listening on 2222. Cloud-init applies the config and restarts sshd.socket, which then tries to bind **again** on 2222. Because the socket lingers, the new socket fails to bind with "Address already in use".

I think the proper solution is to make this a runtime unit.

 ```yaml
#cloud-config

coreos:
  units:
  - name: sshd.socket
    command: restart
    runtime: true
    content: |
      [Socket]
      ListenStream=2222
      Accept=yes
```

crawford commented:
Hmm, I'm having trouble reproducing this failure. @sublimino What version of systemd are you using?

sublimino commented:
Thanks @crawford - FWIW I've not seen this behaviour at first boot.

systemd v218 via alpha channel - currently CoreOS 575.0.0

```
# systemd-run --version
systemd 218
-PAM -AUDIT -SELINUX +IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP -GCRYPT -GNUTLS -ACL -XZ -LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN
```

crawford commented:
That is the version with which I tested. Which cloud provider are you using? Can you walk me through the steps to reproduce?

sublimino commented:
Provider is OVH, dedicated server. I [wrote up my installation](https://www.binarysludge.com/2014/10/02/installing-coreos-on-bare-metal-at-ovh/) a few months back - I still use these exact steps, except I run alpha channel, and with the addition of the `sshd.socket`:

```yaml
#cloud-config

hostname: coreos1
ssh_authorized_keys:
  - ssh-rsa ...
users:
  - name: andy
    groups:
      - sudo
      - docker
    coreos-ssh-import-github: sublimino
coreos:
  etcd:
    name: node001
    discovery: https://discovery.etcd.io/<...>
  fleet:
    metadata: region=<...>
  update:
    reboot-strategy: best-effort
  units:
    - name: sshd.socket
      command: restart
      content: |
        [Socket]
        ListenStream=<in range 49152–65535>
        Accept=yes
        FreeBind=true
    - name: etcd.service
      command: start
    - name: fleet.service
      command: start
    - name: settimezone.service
      command: start
      content: |
        [Unit]
        Description=Set the timezone
        [Service]
        ExecStart=/usr/bin/timedatectl set-timezone UTC
        RemainAfterExit=yes
        Type=oneshot
write_files:
  - path: /etc/resolv.conf
    permissions: 0644
    owner: root
    content: |
      nameserver 8.8.8.8
  - path: /etc/ntp.conf
    content: |
      server 0.pool.ntp.org
      server 1.pool.ntp.org
      restrict default nomodify nopeer noquery limited kod
      restrict 127.0.0.1
      restrict [::1]
```
Apologies for the ~200 line log dump, searching `sshd` reveals useful lines:

```
-- Logs begin at Sat 2015-01-10 20:57:25 UTC, end at Fri 2015-01-30 10:51:29 UTC. --
Jan 24 06:32:51 coreos1 systemd-journal[501]: Runtime journal is using 8.0M (max allowed 3.1G, trying to leave 4.0G free of 31.4G available <E2><86><92> current limit 3.1G).
Jan 24 06:32:51 coreos1 systemd-journal[501]: Permanent journal is using 424.0M (max allowed 4.0G, trying to leave 4.0G free of 1.8T available <E2><86><92> current limit 4.0G).
Jan 24 06:33:00 coreos1 systemd-journal[501]: Time spent on flushing to /var is 8.874066s for 2 entries.
Jan 24 06:33:00 coreos1 systemd-journald[189]: Received SIGTERM from PID 1 (systemd).
Jan 24 06:33:01 coreos1 kernel: igb 0000:03:00.1 enp3s0f1: renamed from eth1
Jan 24 06:33:02 coreos1 kernel: igb 0000:03:00.0 enp3s0f0: renamed from eth0
Jan 24 06:33:02 coreos1 kernel: EXT4-fs (sda6): mounted filesystem with ordered data mode. Opts: commit=600
Jan 24 06:33:02 coreos1 systemd-journal[501]: Journal started
Jan 24 06:32:50 coreos1 systemd[1]: systemd 218 running in system mode. (-PAM -AUDIT -SELINUX +IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP -GCRYPT -GNUTLS -ACL -XZ -LZ4 +SEC
Jan 24 06:32:50 coreos1 systemd[1]: Detected architecture 'x86-64'.
Jan 24 06:32:50 coreos1 systemd[1]: Set hostname to <coreos1>.
Jan 24 06:32:52 coreos1 systemd-udevd[521]: starting version 218
Jan 24 06:32:52 coreos1 systemd-udevd[521]: RUN{builtin}: 'uaccess' unknown /usr/lib64/udev/rules.d/73-seat-late.rules:15
Jan 24 06:33:02 coreos1 systemd[1]: Starting Flush Journal to Persistent Storage...
Jan 24 06:33:08 coreos1 lvm[562]: No volume groups found
Jan 24 06:33:08 coreos1 systemd[1]: Started Flush Journal to Persistent Storage.
Jan 24 06:33:08 coreos1 systemd[1]: Starting Create Volatile Files and Directories...
Jan 24 06:33:09 coreos1 systemd[1]: Started Create Volatile Files and Directories.
Jan 24 06:33:09 coreos1 systemd[1]: Starting Clean up broken links in /etc/ssl/certs...
Jan 24 06:33:09 coreos1 systemd[1]: Starting Update UTMP about System Boot/Shutdown...
Jan 24 06:33:09 coreos1 systemd[1]: Started Clean up broken links in /etc/ssl/certs.
Jan 24 06:33:09 coreos1 systemd[1]: Started Update CA bundle at /etc/ssl/certs/ca-certificates.crt.
Jan 24 06:33:09 coreos1 systemd[1]: Started Update UTMP about System Boot/Shutdown.
Jan 24 06:33:09 coreos1 systemd[1]: Starting System Initialization.
Jan 24 06:33:09 coreos1 systemd[1]: Reached target System Initialization.
Jan 24 06:33:09 coreos1 systemd[1]: Starting Watch for update engine configuration changes.
Jan 24 06:33:09 coreos1 systemd[1]: Listening on sshd.socket.
Jan 24 06:33:09 coreos1 systemd[1]: Listening on Journal Audit Socket.
Jan 24 06:33:09 coreos1 systemd[1]: Starting Sockets.
Jan 24 06:33:10 coreos1 systemd-logind[607]: Watching system buttons on /dev/input/event1 (Power Button)
Jan 24 06:33:11 coreos1 systemd-networkd[599]: lo              : link configured
Jan 24 06:33:11 coreos1 systemd-resolved[628]: Using system hostname 'coreos1'.
Jan 24 06:33:11 coreos1 systemd-logind[607]: Watching system buttons on /dev/input/event0 (Power Button)
Jan 24 06:33:11 coreos1 systemd-udevd[521]: RUN{builtin}: 'uaccess' unknown /usr/lib64/udev/rules.d/73-seat-late.rules:15
Jan 24 06:33:12 coreos1 systemd-logind[607]: New seat seat0.
Jan 24 06:33:12 coreos1 systemd[1]: Started Update Engine.
Jan 24 06:33:12 coreos1 systemd[1]: Starting Cluster reboot manager...
Jan 24 06:33:12 coreos1 systemd[1]: Started Permit User Sessions.
Jan 24 06:33:13 coreos1 kernel: IPv6: ADDRCONF(NETDEV_UP): enp3s0f1: link is not ready
Jan 24 06:33:13 coreos1 kernel: IPv6: ADDRCONF(NETDEV_UP): enp3s0f0: link is not ready
Jan 24 06:33:10 coreos1 ntpd[612]: ntpd 4.2.8@1.3265-o Fri Jan 23 05:33:38 UTC 2015 (1): Starting
Jan 24 06:33:10 coreos1 ntpd[612]: Command line: /usr/sbin/ntpd -g -n -f /var/lib/ntp/ntp.drift -u ntp:ntp
Jan 24 06:33:10 coreos1 ntpd[612]: proto: precision = 0.050 usec (-24)
Jan 24 06:33:10 coreos1 ntpd[612]: Listen and drop on 0 v6wildcard [::]:123
Jan 24 06:33:10 coreos1 ntpd[612]: Listen and drop on 1 v4wildcard 0.0.0.0:123
Jan 24 06:33:10 coreos1 ntpd[612]: Listen normally on 2 lo 127.0.0.1:123
Jan 24 06:33:10 coreos1 ntpd[612]: Listen normally on 3 lo [::1]:123
Jan 24 06:33:10 coreos1 ntpd[612]: Listening on routing socket on fd #20 for interface updates
Jan 24 06:33:11 coreos1 dbus[596]: [system] Activating via systemd: service name='org.freedesktop.hostname1' unit='dbus-org.freedesktop.hostname1.service'
Jan 24 06:33:11 coreos1 dbus[596]: [system] Successfully activated service 'org.freedesktop.hostname1'
Jan 24 06:33:14 coreos1 lvm[568]: No volume groups found
Jan 24 06:33:14 coreos1 lvm[575]: No volume groups found
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063311:INFO:main.cc(159)] CoreOS Update Engine starting
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063311:INFO:payload_state.cc(356)] Current Response Signature =
Jan 24 06:33:14 coreos1 update_engine[601]: NumURLs = 1
Jan 24 06:33:14 coreos1 update_engine[601]: Url0 = https://commondatastorage.googleapis.com/update-storage.core-os.net/amd64-usr/571.0.0/update.gz
Jan 24 06:33:14 coreos1 update_engine[601]: Payload Size = 111137430
Jan 24 06:33:14 coreos1 update_engine[601]: Payload Sha256 Hash = Z2f/y7wNss9oNKTKTRx/HGNWBMTFA2nuUQaO/tNSzZM=
Jan 24 06:33:14 coreos1 update_engine[601]: Is Delta Payload = 0
Jan 24 06:33:14 coreos1 update_engine[601]: Max Failure Count Per Url = 10
Jan 24 06:33:14 coreos1 update_engine[601]: Disable Payload Backoff = 1
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063311:INFO:payload_state.cc(377)] Payload Attempt Number = 1
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063311:INFO:payload_state.cc(400)] Current URL Index = 0
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063311:INFO:payload_state.cc(421)] Current URL (Url0)'s Failure Count = 0
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063311:INFO:payload_state.cc(448)] Backoff Expiry Time = 1/1/1601 0:00:00 GMT
Jan 24 06:33:14 coreos1 update_engine[601]: [0124/063312:INFO:update_check_scheduler.cc(82)] Next update check in 6m17s
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: ntpd 4.2.8@1.3265-o Fri Jan 23 05:33:38 UTC 2015 (1): Starting
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: Command line: /usr/sbin/ntpd -g -n -f /var/lib/ntp/ntp.drift -u ntp:ntp
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: proto: precision = 0.050 usec (-24)
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: Listen and drop on 0 v6wildcard [::]:123
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: Listen and drop on 1 v4wildcard 0.0.0.0:123
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: Listen normally on 2 lo 127.0.0.1:123
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: Listen normally on 3 lo [::1]:123
Jan 24 06:33:14 coreos1 ntpd[612]: 24 Jan 06:33:10 ntpd[612]: Listening on routing socket on fd #20 for interface updates
Jan 24 06:33:14 coreos1 kernel: igb 0000:03:00.0 enp3s0f0: igb: enp3s0f0 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX
Jan 24 06:33:14 coreos1 kernel: IPv6: ADDRCONF(NETDEV_CHANGE): enp3s0f0: link becomes ready
Jan 24 06:33:14 coreos1 systemd-networkd[599]: enp3s0f0        : gained carrier
Jan 24 06:33:14 coreos1 systemd-networkd[599]: enp3s0f0        : DHCPv4 address 5.39.71.195/24 via 5.39.71.254
Jan 24 06:33:14 coreos1 systemd-networkd[599]: enp3s0f0        : link configured
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: Checking availability of "local-file"
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: Fetching user-data from datasource of type "local-file"
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: Fetching meta-data from datasource of type "local-file"
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:10 Parsing user-data as cloud-config
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: Processing cloud-config from user-data
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Set hostname to coreos1
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 User 'andy' exists, ignoring creation-time fields
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Authorizing github user sublimino SSH keys for CoreOS user 'andy'
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Fetching data from https://api.github.com/users/sublimino/keys. Attempt #1
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Unable to fetch data: Get https://api.github.com/users/sublimino/keys: dial tcp: lookup api.github.com: connection
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Sleeping for 100ms...
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Fetching data from https://api.github.com/users/sublimino/keys. Attempt #2
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Unable to fetch data: Get https://api.github.com/users/sublimino/keys: dial tcp: lookup api.github.com: connection
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Sleeping for 200ms...
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Fetching data from https://api.github.com/users/sublimino/keys. Attempt #3
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Unable to fetch data: Get https://api.github.com/users/sublimino/keys: dial tcp: lookup api.github.com: connection
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:11 Sleeping for 400ms...
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:12 Fetching data from https://api.github.com/users/sublimino/keys. Attempt #4
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:12 Unable to fetch data: Get https://api.github.com/users/sublimino/keys: dial tcp: lookup api.github.com: connection
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:12 Sleeping for 800ms...
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:13 Fetching data from https://api.github.com/users/sublimino/keys. Attempt #5
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:13 Unable to fetch data: Get https://api.github.com/users/sublimino/keys: dial tcp: lookup api.github.com: connection
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:13 Sleeping for 1.6s...
Jan 24 06:33:14 coreos1 locksmithd[672]: locksmithd starting currentOperation="UPDATE_STATUS_IDLE" strategy="best-effort"
Jan 24 06:33:14 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:14 Fetching data from https://api.github.com/users/sublimino/keys. Attempt #6
Jan 24 06:33:15 coreos1 ntpd[612]: Listen normally on 4 enp3s0f0 5.39.71.195:123
Jan 24 06:33:15 coreos1 ntpd[612]: bind(24) AF_INET6 fe80::225:90ff:fea6:acbe%2%2#123 flags 0x11 failed: Cannot assign requested address
Jan 24 06:33:15 coreos1 ntpd[612]: unable to create socket on enp3s0f0 (5) for fe80::225:90ff:fea6:acbe%2#123
Jan 24 06:33:15 coreos1 ntpd[612]: failed to init interface for address fe80::225:90ff:fea6:acbe%2
Jan 24 06:33:15 coreos1 ntpd[612]: 24 Jan 06:33:15 ntpd[612]: Listen normally on 4 enp3s0f0 5.39.71.195:123
Jan 24 06:33:15 coreos1 ntpd[612]: 24 Jan 06:33:15 ntpd[612]: bind(24) AF_INET6 fe80::225:90ff:fea6:acbe%2%2#123 flags 0x11 failed: Cannot assign requested address
Jan 24 06:33:15 coreos1 ntpd[612]: 24 Jan 06:33:15 ntpd[612]: unable to create socket on enp3s0f0 (5) for fe80::225:90ff:fea6:acbe%2#123
Jan 24 06:33:15 coreos1 ntpd[612]: 24 Jan 06:33:15 ntpd[612]: failed to init interface for address fe80::225:90ff:fea6:acbe%2
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Authorized SSH keys for core user
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/etc/resolv.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/etc/resolv.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file /etc/resolv.conf to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/etc/ntp.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/etc/ntp.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file /etc/ntp.conf to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/etc/coreos/update.conf"
Jan 24 06:33:15 coreos1 systemd[1]: Starting Generate /run/coreos/motd...
Jan 24 06:33:15 coreos1 systemd[1]: Started Generate /run/coreos/motd.
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/etc/coreos/update.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file /etc/coreos/update.conf to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing unit "sshd.socket" to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/etc/systemd/system/sshd.socket"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/etc/systemd/system/sshd.socket"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote unit "sshd.socket"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing unit "settimezone.service" to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/etc/systemd/system/settimezone.service"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/etc/systemd/system/settimezone.service"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote unit "settimezone.service"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing drop-in unit "20-cloudinit.conf" to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/run/systemd/system/etcd.service.d/20-cloudinit.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/run/systemd/system/etcd.service.d/20-cloudinit.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote drop-in unit "20-cloudinit.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Ensuring runtime unit file "etcd.service" is unmasked
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing drop-in unit "20-cloudinit.conf" to filesystem
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Writing file to "/run/systemd/system/fleet.service.d/20-cloudinit.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote file to "/run/systemd/system/fleet.service.d/20-cloudinit.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Wrote drop-in unit "20-cloudinit.conf"
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Ensuring runtime unit file "fleet.service" is unmasked
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Ensuring runtime unit file "locksmithd.service" is unmasked
Jan 24 06:33:15 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Ensuring runtime unit file "locksmithd.service" is unmasked
Jan 24 06:33:15 coreos1 systemd[1]: Reloading.
Jan 24 06:33:15 coreos1 systemd[1]: Started udev Coldplug all Devices.
Jan 24 06:33:15 coreos1 systemd[1]: Found device HGST_HUS724020ALA640 OEM.
Jan 24 06:33:15 coreos1 systemd[1]: Created slice Root Slice.
Jan 24 06:33:15 coreos1 systemd[1]: Listening on sshd.socket.
Jan 24 06:33:15 coreos1 systemd[1]: Reached target System Initialization.
Jan 24 06:33:15 coreos1 systemd[1]: Started Load/Save Random Seed.
Jan 24 06:33:15 coreos1 systemd[1]: Listening on /dev/initctl Compatibility Named Pipe.
Jan 24 06:33:15 coreos1 systemd[1]: Found device HGST_HUS724020ALA640 OEM.
Jan 24 06:33:15 coreos1 systemd[1]: Found device HGST_HUS724020ALA640 OEM.
Jan 24 06:33:15 coreos1 systemd[1]: Started Rebuild Journal Catalog.
Jan 24 06:33:15 coreos1 systemd[1]: Started Create /etc/nsswitch.conf.
Jan 24 06:33:15 coreos1 systemd[1]: Started Load Kernel Modules.
Jan 24 06:33:15 coreos1 systemd[1]: Starting Clean up broken links in /etc/ssl/certs...
Jan 24 06:33:16 coreos1 systemd[1]: Failed to set memory.limit_in_bytes on : Invalid argument
Jan 24 06:33:16 coreos1 systemd[1]: Failed to reset devices.list on /system.slice: Invalid argument
Jan 24 06:33:16 coreos1 kernel: ip_tables: (C) 2000-2006 Netfilter Core Team
Jan 24 06:33:16 coreos1 systemd[1]: sshd.socket failed to listen on sockets: Address already in use
Jan 24 06:33:16 coreos1 systemd[1]: Unit sshd.socket entered failed state.
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:15 Calling unit command "restart" on "sshd.socket"'
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Result of "restart" on "sshd.socket": failed
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Calling unit command "start" on "etcd.service"'
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Result of "start" on "etcd.service": done
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Calling unit command "start" on "fleet.service"'
Jan 24 06:33:16 coreos1 lvm[786]: No volume groups found
Jan 24 06:33:16 coreos1 lvm[792]: No volume groups found
Jan 24 06:33:16 coreos1 systemd[1]: Started Activation of LVM2 logical volumes.
Jan 24 06:33:16 coreos1 systemd[1]: Time has been changed
Jan 24 06:33:16 coreos1 systemd[1]: Starting fleet.socket.
Jan 24 06:33:16 coreos1 systemd[1]: Listening on fleet.socket.
Jan 24 06:33:16 coreos1 systemd[1]: Starting fleet daemon...
Jan 24 06:33:16 coreos1 systemd[1]: Started fleet daemon.
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Result of "start" on "fleet.service": done
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Calling unit command "start" on "settimezone.service"'
Jan 24 06:33:16 coreos1 systemd[1]: Starting Set the timezone...
Jan 24 06:33:16 coreos1 dbus[596]: [system] Activating via systemd: service name='org.freedesktop.timedate1' unit='dbus-org.freedesktop.timedate1.service'
Jan 24 06:33:16 coreos1 systemd[1]: Starting Time & Date Service...
Jan 24 06:33:16 coreos1 dbus[596]: [system] Successfully activated service 'org.freedesktop.timedate1'
Jan 24 06:33:16 coreos1 systemd[1]: Started Time & Date Service.
Jan 24 06:33:16 coreos1 systemd[1]: Started Set the timezone.
Jan 24 06:33:16 coreos1 systemd[1]: Stopping Cluster reboot manager...
Jan 24 06:33:16 coreos1 systemd[1]: Starting Cluster reboot manager...
Jan 24 06:33:16 coreos1 systemd[1]: Started Cluster reboot manager.
Jan 24 06:33:16 coreos1 systemd[1]: Started Load cloud-config from /var/lib/coreos-install/user_data.
Jan 24 06:33:16 coreos1 systemd[1]: Starting Load user-provided cloud configs.
Jan 24 06:33:16 coreos1 systemd[1]: Reached target Load user-provided cloud configs.
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Result of "start" on "settimezone.service": done
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Calling unit command "restart" on "locksmithd.service"'
Jan 24 06:33:16 coreos1 coreos-cloudinit[613]: 2015/01/24 06:33:16 Result of "restart" on "locksmithd.service": done
Jan 24 06:33:16 coreos1 locksmithd[672]: Received interrupt/termination signal - shutting down.
Jan 24 06:33:16 coreos1 fleetd[798]: INFO fleet.go:58: Starting fleetd version 0.9.0
Jan 24 06:33:16 coreos1 fleetd[798]: INFO fleet.go:162: No provided or default config file found - proceeding without
Jan 24 06:33:17 coreos1 locksmithd[808]: locksmithd starting currentOperation="UPDATE_STATUS_IDLE" strategy="best-effort"
```

crawford commented:
@sublimino I cannot seem to reproduce this. Can you try two things independently for me?
1. Add `ReusePort=yes` to your sshd.socket unit.
2. Make your sshd.socket unit a runtime unit (add `runtime: true` to sshd.socket in your cloud-config).

If my understanding is correct, either one of these should work around the issue.

southskies commented:
I'm on 522.6 stable, no fix here with either `FreeBind=true` or `ReusePort=yes`.

sublimino commented:
I have removed `FreeBind=true ` and run option 1) above for the last few days, but `ReusePort=yes` has been ineffectual: 

```
Feb 05 20:36:37 coreos1 coreos-cloudinit[602]: 2015/02/05 20:36:37 Calling unit command "restart" on "sshd.socket"'
Feb 05 20:36:37 coreos1 systemd[1]: Stopping sshd.socket.
Feb 05 20:36:37 coreos1 systemd[1]: Starting sshd.socket.
Feb 05 20:36:37 coreos1 systemd[1]: sshd.socket failed to listen on sockets: Address already in use
Feb 05 20:36:37 coreos1 systemd[1]: Failed to listen on sshd.socket.
Feb 05 20:36:37 coreos1 systemd[1]: Unit sshd.socket entered failed state.
Feb 05 20:36:37 coreos1 coreos-cloudinit[602]: 2015/02/05 20:36:37 Result of "restart" on "sshd.socket": failed
```

I've found `coreos-cloudinit`'s failure to restart `sshd.socket` tends to occur on all automatic reboots after OS upgrades (on alpha, so frequent).

Now retrying with  
```units:
  - name: sshd.socket
    command: restart
    runtime: true
    content: |
      [Socket]
      ListenStream=<high unreserved port>
      Accept=yes
```

sublimino commented:
`runtime: true` without any other modifications also fails. Post-update the unit fails to start with `sshd.socket failed to listen on sockets: Address already in use`, which means restart fails. Rebooting the machine after this does not have this issue.

```
Feb 13 06:08:22 coreos1 update_engine[586]: [0213/060822:INFO:update_attempter.cc(498)] Update successfully applied, waiting to reboot.
...
Feb 13 06:08:22 coreos1 systemd-logind[591]: System is rebooting.
...
Feb 13 06:10:21 coreos1 systemd[1]: Stopping sshd.socket.
Feb 13 06:10:21 coreos1 systemd[1]: Starting sshd.socket.
Feb 13 06:10:21 coreos1 systemd[1]: sshd.socket failed to listen on sockets: Address already in use
Feb 13 06:10:21 coreos1 systemd[1]: Failed to listen on sshd.socket.
Feb 13 06:10:21 coreos1 systemd[1]: Unit sshd.socket entered failed state.
...
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Writing unit "sshd.socket" to filesystem
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Writing file to "/run/systemd/system/sshd.socket"
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Wrote file to "/run/systemd/system/sshd.socket"
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Wrote unit "sshd.socket"
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Ensuring runtime unit file "sshd.socket" is unmasked
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 /run/systemd/system/sshd.socket is not null or empty, refusing to unmask
...
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Calling unit command "restart" on "sshd.socket"'
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Result of "restart" on "sshd.socket": failed
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Calling unit command "start" on "etcd.service"'
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Result of "start" on "etcd.service": done
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:21 Calling unit command "start" on "fleet.service"'
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:22 Result of "start" on "fleet.service": done
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:22 Calling unit command "start" on "settimezone.service"'
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:22 Result of "start" on "settimezone.service": done
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:22 Calling unit command "restart" on "locksmithd.service"'
Feb 13 06:10:24 coreos1 coreos-cloudinit[597]: 2015/02/13 06:10:22 Result of "restart" on "locksmithd.service": done
```

When I reboot the machine afterwards:

```
Feb 13 07:26:28 coreos1 systemd[1]: Starting Paths.
Feb 13 07:26:28 coreos1 systemd[1]: Reached target Paths.
Feb 13 07:26:28 coreos1 systemd[1]: Starting sshd.socket.
Feb 13 07:26:28 coreos1 systemd[1]: Listening on sshd.socket.
Feb 13 07:26:28 coreos1 systemd[1]: Started Update Engine Stub Timer.
Feb 13 07:26:28 coreos1 systemd[1]: Starting Timers.
...
Feb 13 07:26:28 coreos1 systemd[1]: Starting Generate /run/issue...
Feb 13 07:26:28 coreos1 systemd[1]: Started Modifies /etc/environment for CoreOS.
Feb 13 07:26:28 coreos1 systemd[1]: Starting Generate sshd host keys...
Feb 13 07:26:28 coreos1 systemd[1]: Starting Update Engine...
Feb 13 07:26:28 coreos1 systemd[1]: Starting Login Service...
...
Feb 13 07:26:29 coreos1 kernel: IPv6: ADDRCONF(NETDEV_UP): enp3s0f0: link is not ready
Feb 13 07:26:29 coreos1 systemd[1]: Started Network Name Resolution.
Feb 13 07:26:29 coreos1 systemd[1]: Started Generate sshd host keys.
Feb 13 07:26:29 coreos1 systemd-udevd[516]: RUN{builtin}: 'uaccess' unknown /usr/lib64/udev/rules.d/73-seat-late.rules:15
Feb 13 07:26:30 coreos1 dbus[596]: [system] Successfully activated service 'org.freedesktop.hostname1'
...
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Wrote file to "/etc/coreos/update.conf"
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Wrote file /etc/coreos/update.conf to filesystem
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Writing unit "sshd.socket" to filesystem
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Writing file to "/run/systemd/system/sshd.socket"
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Wrote file to "/run/systemd/system/sshd.socket"
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Wrote unit "sshd.socket"
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Ensuring runtime unit file "sshd.socket" is unmasked
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 /run/systemd/system/sshd.socket is not null or empty, refusing to unmask
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Writing unit "settimezone.service" to filesystem
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Writing file to "/etc/systemd/system/settimezone.service"
...
Feb 13 07:26:37 coreos1 systemd[1]: Started Run Addon /usr/share/oem.
Feb 13 07:26:37 coreos1 systemd[1]: Listening on Journal Audit Socket.
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Calling unit command "restart" on "sshd.socket"'
Feb 13 07:26:37 coreos1 systemd[1]: Stopping sshd.socket.
Feb 13 07:26:37 coreos1 systemd[1]: Starting sshd.socket.
Feb 13 07:26:37 coreos1 systemd[1]: Listening on sshd.socket.
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Result of "restart" on "sshd.socket": done
Feb 13 07:26:37 coreos1 coreos-cloudinit[594]: 2015/02/13 07:26:37 Calling unit command "start" on "etcd.service"'
Feb 13 07:26:37 coreos1 systemd[1]: Starting etcd...
...
```

Perhaps a combination of `ReusePort=yes` and `runtime: true`?

sublimino commented:
A combination of `ReusePort=yes` and `runtime: true` seems to have solved the failure-to-restart problem on the alpha channel.

sublimino commented:
I've had a couple more successful unmanaged reboots. Here is the relevant `cloud-config` excerpt:

```
  - name: sshd.socket
    command: restart
    runtime: true
    content: |
      [Socket]
      ListenStream=22222
      Accept=yes
      ReusePort=yes
```

Happy to try further/repeat previous combinations if the above is not what is expected.

sublimino commented:
This issue repeated three times between ~5th March and 23rd April upon Alpha channel updates (of which [there were 7](https://coreos.com/releases/#653.0.0))

osterman commented:
I'm still having this problem daily.  Adding `ReusePort=yes` and `runtime: true` has not made any difference for me on `CoreOS 681.0.0`.

osterman commented:
Here's some more information. 

Session transcript of `sshd.socket`:

    Jun 11 02:07:26 core2.cluster-2.cloudposse.com systemd[1]: Closed sshd.socket.
    Jun 11 02:07:26 core2.cluster-2.cloudposse.com systemd[1]: Stopping sshd.socket.
    -- Reboot --
    Jun 11 02:07:48 core2.cluster-2.cloudposse.com systemd[1]: Listening on sshd.socket.
    Jun 11 02:07:48 core2.cluster-2.cloudposse.com systemd[1]: Starting sshd.socket.
    Jun 11 02:07:53 core2.cluster-2.cloudposse.com systemd[1]: Stopping sshd.socket.
    Jun 11 02:07:53 core2.cluster-2.cloudposse.com systemd[1]: sshd.socket failed to listen on sockets: Address already in use
    Jun 11 02:07:53 core2.cluster-2.cloudposse.com systemd[1]: Failed to listen on sshd.socket.
    Jun 11 02:07:53 core2.cluster-2.cloudposse.com systemd[1]: Unit sshd.socket entered failed state.
    Jun 11 02:07:53 core2.cluster-2.cloudposse.com systemd[1]: Starting sshd.socket.

I was sound asleep as the events above transpired. Not sure why it went from Starting->Stoping->Address in use->Failed->Starting, but still no working socket.

To recover the service so that I could connect to the machine, I had to use fleet to trigger a unit that calls “sudo systemctl restart sshd.socket”. 
   
    Jun 11 03:04:28 core2.cluster-2.cloudposse.com systemd[1]: Listening on sshd.socket.
    Jun 11 03:04:28 core2.cluster-2.cloudposse.com systemd[1]: Starting sshd.socket.

I will use this as my stop gap measure since this happens somewhat regularly.

This is all despite having the socket configured with the suggested options (both `runtime: true` and `ReusePort=yes`). 

    core2 ~ # cat /run/systemd/system/sshd.socket
    [Socket]
    ListenStream=48622
    Accept=yes
    ReusePort=yes

joshix commented:
I believe this has been picked up and covered in https://github.com/coreos/docs/blob/master/os/customizing-sshd.md#changing-the-sshd-port-1
